File: packages/python/procedurals/binary_cell_partition.py
"""Immutable ordered integer-cell rectangles from attempt-bounded binary cuts.

Implements layout.binary-cell-partition-2d 0.1.0. Independently specified from
survey/out/2018/Generativos/poop/notes.md and survey/out/2018/Generativos/barab/notes.md.
CP17's private native study observed attempts 20/80/240 and RANDOM/LONGEST policies;
these are not defaults or encouraged ranges.
"""
import math

__all__ = ["PartitionError", "binary_cell_partition_2d"]

_MAX_ATTEMPTS = 2147483646
_UINT32_MAX = 4294967295
_INT32_MAX = 2147483647
_KEYS = ("seed", "columns", "rows", "attempts", "axisPolicy")
_MASK32 = (1 << 32) - 1
_MASK64 = (1 << 64) - 1


class PartitionError(ValueError):
    """Stable catalog code for binary-cell-partition-2d."""
    def __init__(self, code):
        super().__init__(code)
        self.code = code


def _invalid():
    raise PartitionError("INVALID_INPUT")


def _integer(value, minimum, maximum):
    if type(value) not in (int, float):
        _invalid()
    if type(value) is float and (not math.isfinite(value) or not value.is_integer()):
        _invalid()
    if not minimum <= value <= maximum:
        _invalid()
    return int(value)


def _record(value):
    if type(value) is not dict:
        _invalid()
    if len(value) != len(_KEYS) or any(key not in value for key in _KEYS):
        _invalid()
    return value


class _Stream:
    """Private xoshiro128**1.1 stream with two SplitMix64 expansion words.

    Independently duplicated from quadrant_partition.py's _Stream per the established
    per-module convention; both implement the shared private RNG contract.
    """
    def __init__(self, seed):
        x = seed + 0x9E3779B97F4A7C15
        a = self._split(x)
        b = self._split(x + 0x9E3779B97F4A7C15)
        self._s = [a & _MASK32, a >> 32, b & _MASK32, b >> 32]

    @staticmethod
    def _split(x):
        x &= _MASK64
        x = ((x ^ (x >> 30)) * 0xBF58476D1CE4E5B9) & _MASK64
        x = ((x ^ (x >> 27)) * 0x94D049BB133111EB) & _MASK64
        return (x ^ (x >> 31)) & _MASK64

    def output(self):
        s = self._s
        value = (s[1] * 5) & _MASK32
        result = ((((value << 7) | (value >> 25)) & _MASK32) * 9) & _MASK32
        t = (s[1] << 9) & _MASK32
        s[2] ^= s[0]
        s[3] ^= s[1]
        s[1] ^= s[2]
        s[0] ^= s[3]
        s[2] ^= t
        s[3] = ((s[3] << 11) | (s[3] >> 21)) & _MASK32
        return result

    def unit(self):
        return self.output() / 4294967296.0


class _Partition:
    __slots__ = ("_bottom", "_left", "_right", "_splits", "_top")

    def __init__(self, left, top, right, bottom, splits):
        self._left = left
        self._top = top
        self._right = right
        self._bottom = bottom
        self._splits = splits

    @property
    def size(self):
        return len(self._left)

    @property
    def splits(self):
        return self._splits

    def _checked_index(self, index):
        if type(index) not in (int, float):
            raise PartitionError("INVALID_INDEX")
        if type(index) is float and (not math.isfinite(index) or not index.is_integer()):
            raise PartitionError("INVALID_INDEX")
        if not 0 <= index <= 9007199254740991:
            raise PartitionError("INVALID_INDEX")
        index = int(index)
        if index >= self.size:
            raise PartitionError("INDEX_OUT_OF_RANGE")
        return index

    def bounds_at(self, index):
        i = self._checked_index(index)
        return [self._left[i], self._top[i], self._right[i], self._bottom[i]]

    def bounds_into(self, index, destination):
        i = self._checked_index(index)
        if type(destination) is not list or len(destination) != 4:
            raise PartitionError("INVALID_OUTPUT")
        destination[0] = self._left[i]
        destination[1] = self._top[i]
        destination[2] = self._right[i]
        destination[3] = self._bottom[i]
        return destination

    def to_values(self):
        bounds = [[self._left[i], self._top[i], self._right[i], self._bottom[i]] for i in range(self.size)]
        return {"bounds": bounds, "splits": self._splits}


def binary_cell_partition_2d(config):
    """Generate a tiling of [0,0,columns,rows] using a private uint32-seeded stream."""
    _record(config)
    seed = _integer(config["seed"], 0, _UINT32_MAX)
    columns = _integer(config["columns"], 1, _INT32_MAX)
    rows = _integer(config["rows"], 1, _INT32_MAX)
    attempts = _integer(config["attempts"], 0, _MAX_ATTEMPTS)
    axis_policy = config["axisPolicy"]
    if axis_policy not in ("RANDOM", "LONGEST"):
        _invalid()
    random_axis = axis_policy == "RANDOM"

    left, top, right, bottom = [0], [0], [columns], [rows]
    size, splits = 1, 0
    stream = _Stream(seed)
    for _ in range(attempts):
        selected = math.floor(stream.unit() * size)
        x, y = left[selected], top[selected]
        x2, y2 = right[selected], bottom[selected]
        width, height = x2 - x, y2 - y
        split_width = (stream.unit() < 0.5) if random_axis else (width > height)
        extent = width if split_width else height
        if extent == 1:
            continue
        cut = 1 + math.floor(stream.unit() * (extent - 1))

        survivors_left = left[selected + 1:size]
        survivors_top = top[selected + 1:size]
        survivors_right = right[selected + 1:size]
        survivors_bottom = bottom[selected + 1:size]
        del left[selected:], top[selected:], right[selected:], bottom[selected:]
        left.extend(survivors_left); top.extend(survivors_top)
        right.extend(survivors_right); bottom.extend(survivors_bottom)

        left.extend([x, x + cut if split_width else x])
        top.extend([y, y if split_width else y + cut])
        right.extend([x + cut if split_width else x2, x2])
        bottom.extend([y2 if split_width else y + cut, y2])
        size += 1
        splits += 1
    return _Partition(left, top, right, bottom, splits)

File: packages/javascript/src/binary-cell-partition.js
const MAX_ATTEMPTS = 2147483646;
const UINT32_MAX = 4294967295;
const INT32_MAX = 2147483647;
const KEYS = ["seed", "columns", "rows", "attempts", "axisPolicy"];

/** Validation/access error with a stable catalog code. */
export class PartitionError extends Error {
  constructor(code) {
    super(code);
    this.name = "PartitionError";
    this.code = code;
  }
}

function invalid() { throw new PartitionError("INVALID_INPUT"); }

function integer(value, minimum, maximum) {
  if (typeof value !== "number" || !Number.isFinite(value)) invalid();
  if (!Number.isInteger(value) || value < minimum || value > maximum) invalid();
  return value;
}

function record(value) {
  if (value === null || typeof value !== "object" || Array.isArray(value)) invalid();
  const prototype = Object.getPrototypeOf(value);
  if (prototype !== Object.prototype && prototype !== null) invalid();
  if (Reflect.ownKeys(value).length !== KEYS.length) invalid();
  for (const key of KEYS) if (!Object.prototype.hasOwnProperty.call(value, key)) invalid();
  return value;
}

function writableSlot(array, index) {
  const key = String(index);
  const descriptor = Object.getOwnPropertyDescriptor(array, key);
  if (descriptor !== undefined) return "value" in descriptor && descriptor.writable === true;
  if (!Object.isExtensible(array)) return false;
  for (let prototype = Object.getPrototypeOf(array); prototype !== null; prototype = Object.getPrototypeOf(prototype)) {
    if (Object.getOwnPropertyDescriptor(prototype, key) !== undefined) return false;
  }
  return true;
}

// Private xoshiro128**1.1 stream, expanded through two SplitMix64 outputs.
// Identical to quadrant-partition.js's Stream (independently duplicated per the
// established per-module convention; both implement the shared private RNG contract).
class Stream {
  constructor(seed) {
    let x = BigInt(seed) + 0x9e3779b97f4a7c15n;
    const a = this.split(x);
    x += 0x9e3779b97f4a7c15n;
    const b = this.split(x);
    this.s0 = Number(a & 0xffffffffn) >>> 0;
    this.s1 = Number(a >> 32n) >>> 0;
    this.s2 = Number(b & 0xffffffffn) >>> 0;
    this.s3 = Number(b >> 32n) >>> 0;
  }
  split(x) {
    x = BigInt.asUintN(64, x);
    x = BigInt.asUintN(64, (x ^ (x >> 30n)) * 0xbf58476d1ce4e5b9n);
    x = BigInt.asUintN(64, (x ^ (x >> 27n)) * 0x94d049bb133111ebn);
    return BigInt.asUintN(64, x ^ (x >> 31n));
  }
  output() {
    const value = Math.imul(this.s1, 5) >>> 0;
    const result = Math.imul(((value << 7) | (value >>> 25)) >>> 0, 9) >>> 0;
    const t = (this.s1 << 9) >>> 0;
    this.s2 = (this.s2 ^ this.s0) >>> 0;
    this.s3 = (this.s3 ^ this.s1) >>> 0;
    this.s1 = (this.s1 ^ this.s2) >>> 0;
    this.s0 = (this.s0 ^ this.s3) >>> 0;
    this.s2 = (this.s2 ^ t) >>> 0;
    this.s3 = ((this.s3 << 11) | (this.s3 >>> 21)) >>> 0;
    return result;
  }
  unit() { return this.output() / 4294967296; }
}

class Partition {
  #left; #top; #right; #bottom; #splits;
  constructor(left, top, right, bottom, splits) {
    this.#left = left; this.#top = top; this.#right = right; this.#bottom = bottom;
    this.#splits = splits;
    Object.freeze(this);
  }
  get size() { return this.#left.length; }
  get splits() { return this.#splits; }
  #checkedIndex(index) {
    if (typeof index !== "number" || !Number.isInteger(index) || index < 0 || index > 9007199254740991) {
      throw new PartitionError("INVALID_INDEX");
    }
    if (index >= this.size) throw new PartitionError("INDEX_OUT_OF_RANGE");
    return index;
  }
  boundsAt(index) {
    const i = this.#checkedIndex(index);
    return [this.#left[i], this.#top[i], this.#right[i], this.#bottom[i]];
  }
  boundsInto(index, destination) {
    const i = this.#checkedIndex(index);
    if ((!Array.isArray(destination) && !(destination instanceof Int32Array)) || destination.length !== 4 ||
        !writableSlot(destination, 0) || !writableSlot(destination, 1) ||
        !writableSlot(destination, 2) || !writableSlot(destination, 3)) {
      throw new PartitionError("INVALID_OUTPUT");
    }
    destination[0] = this.#left[i]; destination[1] = this.#top[i];
    destination[2] = this.#right[i]; destination[3] = this.#bottom[i];
    return destination;
  }
  toValues() {
    const bounds = new Array(this.size);
    for (let i = 0; i < this.size; i += 1) bounds[i] = [this.#left[i], this.#top[i], this.#right[i], this.#bottom[i]];
    return { bounds, splits: this.#splits };
  }
}

/**
 * Generate immutable ordered integer-cell rectangles from attempt-bounded binary cuts.
 * Implements layout.binary-cell-partition-2d 0.1.0.
 *
 * Independently specified from survey/out/2018/Generativos/poop/notes.md and
 * survey/out/2018/Generativos/barab/notes.md. CP17's private native study observed
 * attempts 20/80/240 and RANDOM/LONGEST policies; these are not defaults or
 * encouraged ranges.
 */
export function binaryCellPartition2D(input) {
  record(input);
  const seed = integer(input.seed, 0, UINT32_MAX);
  const columns = integer(input.columns, 1, INT32_MAX);
  const rows = integer(input.rows, 1, INT32_MAX);
  const attempts = integer(input.attempts, 0, MAX_ATTEMPTS);
  const axisPolicy = input.axisPolicy;
  if (axisPolicy !== "RANDOM" && axisPolicy !== "LONGEST") invalid();
  const randomAxis = axisPolicy === "RANDOM";

  let left = [0], top = [0], right = [columns], bottom = [rows];
  let size = 1, splits = 0;
  const stream = new Stream(seed);
  for (let attempt = 0; attempt < attempts; attempt += 1) {
    const selected = Math.floor(stream.unit() * size);
    const x = left[selected], y = top[selected];
    const x2 = right[selected], y2 = bottom[selected];
    const width = x2 - x, height = y2 - y;
    const splitWidth = randomAxis ? stream.unit() < 0.5 : width > height;
    const extent = splitWidth ? width : height;
    if (extent === 1) continue;
    const cut = 1 + Math.floor(stream.unit() * (extent - 1));

    const survivors = left.slice(selected + 1, size);
    const survivorsTop = top.slice(selected + 1, size);
    const survivorsRight = right.slice(selected + 1, size);
    const survivorsBottom = bottom.slice(selected + 1, size);
    left.length = selected; top.length = selected; right.length = selected; bottom.length = selected;
    left.push(...survivors); top.push(...survivorsTop); right.push(...survivorsRight); bottom.push(...survivorsBottom);

    left.push(x, splitWidth ? x + cut : x);
    top.push(y, splitWidth ? y : y + cut);
    right.push(splitWidth ? x + cut : x2, x2);
    bottom.push(splitWidth ? y2 : y + cut, y2);
    size += 1;
    splits += 1;
  }
  return new Partition(left, top, right, bottom, splits);
}
